// @native-html/plugins-core/lib/commonjs/linkPressTargetToOnDOMLinkPressArgs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = linkPressTargetToOnDOMLinkPressArgs;

/**
 * This function converts an "click event" from webshell to an array of
 * arguments for `onLinkPress` callback from react-native-render-html.
 *
 * @param param0 - The event from webshell
 * @public
 */
function linkPressTargetToOnDOMLinkPressArgs({
  uri,
  classAttribute,
  downloadAttribute,
  hrefAttribute,
  hreflangAttribute,
  targetAttribute,
  idAttribute,
  nameAttribute,
  referrerpolicyAttribute,
  relAttribute,
  typeAttribute
}) {
  const attributes = {
    class: classAttribute,
    download: downloadAttribute,
    href: hrefAttribute,
    hreflang: hreflangAttribute,
    id: idAttribute,
    name: nameAttribute,
    referrerpolicy: referrerpolicyAttribute,
    rel: relAttribute,
    type: typeAttribute,
    target: targetAttribute
  };

  for (const name in attributes) {
    if (attributes[name] == null) {
      delete attributes[name];
    }
  }

  return [{
    nativeEvent: {}
  }, uri, attributes, targetAttribute || '_self'];
}
